plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,300 @@
+import java.util.Map;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting
+ * top N queries with high latency data for search queries
+ *
+ * @opensearch.internal
+ */
+public class QueryInsightsService {
+    private static final Logger log = LogManager.getLogger(QueryInsightsService.class);
+
+    private static final TimeValue delay = TimeValue.ZERO;
+    /**
+     * The internal OpenSearch thread pool that execute async processing and exporting tasks
+     */
+    private final ThreadPool threadPool;
+
+    /**
+     * enable insight data collection
+     */
+    private final Map<MetricType, Boolean> enableCollect = new HashMap<>();
+
+    private int topNSize = QueryInsightsSettings.DEFAULT_TOP_N_SIZE;
+
+    private TimeValue windowSize = TimeValue.timeValueSeconds(QueryInsightsSettings.DEFAULT_WINDOW_SIZE);
+    /**
+     * The internal thread-safe store that holds the top n queries insight data, by different MetricType
+     */
+    private final Map<MetricType, PriorityBlockingQueue<SearchQueryRecord>> topQueriesStores;
+
+    /**
+     * The internal store that holds historical top n queries insight data by different MetricType in the last window
+     */
+    private final Map<MetricType, List<SearchQueryRecord>> topQueriesHistoryStores;
+    /**
+     * window start timestamp for each top queries collector
+     */
+    private final Map<MetricType, Long> topQueriesWindowStart = new HashMap<>();
+
+    /**
+     * Create the TopQueriesByLatencyService Object
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool) {
+        topQueriesStores = new HashMap<>();
+        topQueriesHistoryStores = new HashMap<>();
+        for (MetricType metricType : MetricType.allMetricTypes()) {
+            topQueriesStores.put(metricType, new PriorityBlockingQueue<>(topNSize, (a, b) -> SearchQueryRecord.compare(a, b, metricType)));
+            topQueriesHistoryStores.put(metricType, new ArrayList<>());
+            topQueriesWindowStart.put(metricType, -1L);
+        }
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Ingest the query data into local stores
+     *
+     * @param record the record to ingest
+     */
+    public void addRecord(SearchQueryRecord record) {
+        Map<MetricType, Measurement<? extends Number>> measurements = record.getMeasurements();
+        for (MetricType metricType : measurements.keySet()) {
+            if (!topQueriesStores.containsKey(metricType)) {
+                continue;
+            }
+            // add the record to corresponding priority queues to calculate top n queries insights
+            PriorityBlockingQueue<SearchQueryRecord> store = topQueriesStores.get(metricType);
+            this.threadPool.schedule(() -> {
```

Comment:
It would be nice if you can keep the stats of query insights plugin. Stats can include - `number_of_failures`, `total_processed` and general information about the health of the plugin. You can plug this logic as async action listener here to collect them. If it makes sense please create a issue with list of things you are planning to do as part of 2.13.

Reply:
make sense. Let me create issues for those items.